docs/design/data_model.md

```diff
@@ -0,0 +1,412 @@
+# Holochain Data Model Design
+
+## Overview
+
+This document describes the data model used by Holochain for representing and persisting application data.
+
+## Core Concepts
+
+### Agent
+
+An _Agent_ is any entity which can exercise the property of agency within a Holochain network. In practice, this refers to a participant in a distributed application who can author data, validate the actions of others, and participate in the peer-to-peer network.
+
+Each agent is identified by an _AgentPubKey_, which is the public half of a cryptographic signing key pair. While called a "key", it functions as an identity hash within the system.
+
+### Action
+
+An _Action_ is the fundamental unit of authorship in Holochain. Actions form a hash chain (a.k.a "source chain") that provides an immutable, tamper-evident record of everything an agent has done. Actions come in several types, each of which carries semantic meaning.
+
+### Entry
+
+An _Entry_ is application data that can be stored in Holochain. Entries represent the actual content that applications work with, while actions are defined by the system and 
+describe how the entry data should be interpreted.
+
+### Record
+
+A _Record_ is the combination of an action, signed by the agent who authored it, together with its associated entry if applicable. Records are the primary unit of data that applications query and work with.
+
+The entry part of a record exists in one of multiple states. An entry may be _present_ when the entry data is included with the action. It may be _hidden_ when the entry exists but is marked as private and therefore not shared. Some action types don't permit an entry at all, in which case the entry is _not applicable_. Finally, an action may be stored without its entry by agents other than the author, resulting in a _not stored_ state.
+
+### Chain DHT Operations
+
+Represent chain data being shared with the network. Each action generates a specific set of chain operations (such as _StoreRecord_, _RegisterAgentActivity_, _StoreEntry_, etc.) that enable different authorities to store and index the data for efficient queries.
+
+### Warrant Operations
+
+Represent evidence of invalid behavior by an agent. When an agent detects that another agent has violated validation rules or forked their chain, they can issue a warrant containing cryptographic proof of the violation. Warrants are routed to the violating agent's activity authority, creating a permanent, queryable record of the misbehavior.
+
+### Location
+
+A location is a derived property of a hash. The exact algorithm is not important, as long as it preserves the uniqueness of its hash and all peers are using the same algorithm. The location value is an unsigned 32-bit number.
```

Comment:
"preserves the uniqueness of the hash" is a bit unclear language since we're reducing the 32-byte hash to 32-bits thereby significantly reducing its uniqueness.